posix: add pthread_key

nniranjhana opened the pull request:
This patch does the following:

(i) Add a pthread key object and key data structures,
add a key list to the posix_struct for threads.
(ii) Add 4 new pthread_key APIs for thread-specific data key
creation, deletion, setting and retrieving the values.
(iii) Add pthread_once API and pthread_once_t variables used
for dynamic package initialization.
(iv) Add a simple test case which tests all the implemented APIs.

Signed-off-by: Niranjhana N <niranjhana.n@intel.com>

Commit:
posix: add pthread_key and pthread_once APIs

Added 4 new pthread_key APIs for thread-specific data
key creation, deletion, setting and getting the values.

Added a key list to the posix_struct for threads.

Added pthread_once API.

Signed-off-by: Niranjhana N <niranjhana.n@intel.com>

lpereira commented on include/posix/pthread.h:
```diff
@@ -41,6 +41,9 @@ struct posix_thread {
 	pthread_cond_t state_cond;
+
+	/* List of keys that thread has called pthread_setspecific() on */
+	sys_dlist_t keyList;
```
Name this `key_list`.  thereIsNoCamelCaseInZephyr.

lpereira commented on include/posix/sys/types.h:
```diff
@@ -76,6 +76,39 @@ typedef struct pthread_rwlock_obj {
+	sys_dlist_t KeyDataL;
+} pthread_key_obj;
+
+typedef struct ThreadData {
```
This is an implementation detail, and as such, this should be declared in the implementation file only.  The rationale is to not pollute the namespace of any file including `sys/types.h` with things that are not there by the standard.  Also, please pay attention to the naming scheme used throughout the project; use `unix_hacker_style` instead of `CamelCaseStyle`.

nniranjhana replied:
Okay, I have moved both the `thread_data` and `key_data` structs to the implementation file `pthread_key.c`, but retained the `pthread_key_obj` struct in the `sys/types.h` file only.

I have also renamed these (to follow `unix_hacker_style` convention) to start with `pthread_` so as to avoid conflicts with the corresponding type structure objects declared and used in the pthread_key APIs.

lpereira commented on include/posix/sys/types.h:
```diff
@@ -76,6 +76,39 @@ typedef struct pthread_rwlock_obj {
+	/* Back pointer for easy removal of thread data from KeyDataL,
+	 * the key's thread data list, when the thread exits
+	 */
+	struct KeyData *keyData;
```
It might be OK to not have this pointer and traverse `KeyDataL` when thread exits.  Threads exiting are not that common, and this saves a pointer from this struct.

nniranjhana replied:
Okay, I have removed it here, and also in `pthread_setspecific`, where it is initialized.

lpereira commented on lib/posix/pthread.c:
```diff
@@ -302,6 +302,21 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+ */
+int pthread_once(pthread_once_t *once, void (*initFunc)(void))
+{
+	if (*once == PTHREAD_ONCE_INIT) {
```
This is racy: please use `atomic.h` here.  What you want is a `compare-and-set` primitive (`atomic_cas()` in `atomic.h`):  if you successfully compare `*once` with, say, a "not initialized" value, swap it with a "initialized value" and call the init function.  It is close to what you're doing here, but this operation must be atomic; otherwise, `initFunc()` may be called twice if this code is preempted midway.

lpereira replied:
This might work:

```C
#define NEVER 0
#define INITIALIZING 1
#define INITIALIZED 2

void once(atomic_val_t *once, void (*fn)(void)) {
  /* Nothing to do: fn() has been called once according to the *once
   * value. */
  if (atomic_get(once) == INITIALIZED) {
      return;
  }

  /* Try going from "never ran" to "will run"... */
  if (atomic_cas(once, NEVER, INITIALIZING)) {
      fn();
      /* ...and then go to "we executed" once fn() has executed */
      atomic_inc(once);
      return;
  }

  /* atomic_cas() returned false; in this case, it's in either
   * INITIALIZING or INITIALIZED at this point.  Spin on it
   * until it changes its value to INITIALIZED.
   */
  while (atomic_get(once) != INITIALIZED) {
      k_yield();
  }
}
```

lpereira replied:
@andyross Could you take a look here?  Do you see anything wrong with this `once()` implementation above?

andyross replied:
That would work, but spinning on yield is just a bad idea.  The implication from way it's documented is that any blocking will be OS-managed.   You can do this with a condition variable relatively easily, just implement this in terms of other POSIX APIs..  Put a pthread_mutex around the "once" accesses, and in the loop at the end do ptherad_cond_wait() instead of yield, with a pthread_cond_broadcast() after fn() has returned.

nniranjhana replied:
@lpereira Thanks a lot for explaining this in good detail, I can see the importance of handling race condition now. Shall I go ahead and use POSIX APIs to implement this, like Andy has mentioned?

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,172 @@
+
+	if (newKey == NULL) {
+		errno = ENOMEM;
+		return -1;
```
This function does not set `errno`, but rather return the error code directly:

```
RETURN VALUE
       If  successful,  the pthread_key_create() function shall store the newly created key value
       at *key and shall return zero. Otherwise, an error number shall be  returned  to  indicate
       the error.

```

nniranjhana replied:
Yes, you are right. I will change this accordingly to return the error codes directly in other APIs as well

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,172 @@
+	pthread_key_obj *keyObj = (pthread_key_obj *)key;
+	KeyData *keyData;
+
+	sem_wait(&sem);
```
For the POSIX implementation, it might be a good idea to use Zephyr objects directly, to reduce the overhead.

nniranjhana replied:
Okay, I have changed this to use Zephyr `k_sem` objects directly

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,172 @@
+	while (!sys_dlist_is_empty(&(keyObj->KeyDataL))) {
+
+		/* Remove the object from the list KeyDataL */
+		keyData = (KeyData *)sys_dlist_get(&(keyObj->KeyDataL));
```
Casting from a `void *` to any other pointer in C without a cast is valid.  Please remove all casts that are unnecessary.

nniranjhana replied:
Okay, if I remove them, they work, but here, `sys_dlist_get()`'s return type is `sys_dnode_t *`, and not `void *`, so when compiling, it complains of a `incompatible pointer types`. Do I still remove it?

nniranjhana replied:
I will remove the cast in other places, like in `k_malloc` calls

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,172 @@
+	sem_wait(&sem);
+
+	/* Delete thread-specific elements associated with the key */
+	while (!sys_dlist_is_empty(&(keyObj->KeyDataL))) {
```
Also, while it's fine to pass `NULL` to `k_free()`, but since `sys_dlist_get()` will call `sys_dlist_is_empty()` before attempting to dereference `list`, you might want to use `SYS_DLIST_FOR_EACH_NODE_SAFE`; for instance, take this example from `mem_domain.c`:

```C
        SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
                struct k_thread *thread =
                        CONTAINER_OF(node, struct k_thread, mem_domain_info);

                sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
                thread->mem_domain_info.mem_domain = NULL;
        }
```

nniranjhana replied:
Oh alright, I will use this macro instead to loop over the list elements.

So then my loop would be

    SYS_DLIST_FOR_EACH_NODE_SAFE(&(key_obj->key_data_l), node_l, node_next_l) {
        key_data = (pthread_key_data *)sys_dlist_get(&(key_obj->key_data_l));
        k_free((void *)key_data);
    }
 
I would still call `sys_dlist_get()` though, to get the node, so I can free its memory. Is that okay? Or should I directly call `sys_dlist_remove()`?

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,172 @@
+
+	nodeL = sys_dlist_peek_head(&(thread->keyList));
+
+	if (nodeL != NULL) {
```
There are macros to traverse dlists; please use them instead of using open-coded loops such as these.

nniranjhana replied:
Yeah, this is quite messy I agree, I have replaced these four line conditions in the `if` and `do-while` with a single line macro `SYS_DLIST_FOR_EACH_NODE_SAFE` to loop over the list elements. The test passes with all these modifications, do take a look to see if there's something I have missed.

lpereira reviewed (changes requested):
I have not thoroughly reviewed this patch, but there are a few comments that you will want to address before this gets merged.

nniranjhana commented:
@lpereira thank you for reviewing this with good detail, I will address them one by one and update here

ramakrishnapallala commented on include/posix/pthread.h:
```diff
@@ -41,6 +41,9 @@ struct posix_thread {
 	pthread_cond_t state_cond;
+
+	/* List of keys that thread has called pthread_setspecific() on */
+	sys_dlist_t key_list;
```
move to the begining of the struct

nniranjhana replied:
done x2

ramakrishnapallala commented on include/posix/sys/types.h:
```diff
@@ -76,6 +76,21 @@ typedef struct pthread_rwlock_obj {
+	/* List of pthread_key_data objects that contain thread
+	 * specific data for the key
+	 */
+	sys_dlist_t key_data_l;
```
move to the beginning of struct

ramakrishnapallala commented on include/posix/pthread.h:
```diff
@@ -52,6 +55,9 @@ struct posix_thread {
 #define PTHREAD_CANCEL_DISABLE	(1 << _PTHREAD_CANCEL_POS)
 
+/* Passed to pthread_once */
+#define PTHREAD_ONCE_INIT 2
```
this is not needed, as we need to increment value only once.

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -302,6 +309,35 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+		return 0;
+	}
+
+	if (atomic_cas(once, NOT_INITIALIZED, INITIALIZING)) {
```
do only compare here and in the block increment.

nniranjhana replied:
okay, implemented it this way

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -302,6 +309,35 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+		pthread_mutex_lock(&lock);
+		(*initFunc)();
+		pthread_cond_broadcast(&cvar0);
+		atomic_inc(once);
```
initialize it before broadcast

nniranjhana replied:
okay

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -16,6 +16,13 @@
 #define LOWEST_POSIX_THREAD_PRIORITY 1
 
+#define NOT_INITIALIZED 0
+#define INITIALIZING 1
```
PTHREAD_ONCE_NOT_INITIALIZED 0
PTHREAD_ONCE_INITIALIZED 1

ramakrishnapallala commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,181 @@
+#include <misc/dlist.h>
+#include <posix/pthread.h>
+
+struct k_sem sem;
```
give pthread_key_sem or something similar

nniranjhana replied:
done

nniranjhana replied:
done

ramakrishnapallala commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,181 @@
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
```
*key?

ramakrishnapallala commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,181 @@
+		/* Deallocate the object's memory */
+		k_free((void *)key_data);
+	}
+
```
call destructor here?

nniranjhana replied:
will leave this for now, not sure if the thread setting key would be calling it at exit

nniranjhana replied:
I have added the destructor call in `pthread_exit`

ramakrishnapallala reviewed (commented):
please fix the comments

nniranjhana commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,181 @@
+	node_l = sys_dlist_peek_head(&(thread->key_list));
+
+	/* Traverse the list of keys set by the thread, looking for key */
+	while (node_l != (sys_dnode_t *)(&thread->key_list)) {
```
okay

nniranjhana commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,181 @@
+			}
+	}
+
+	if (!found || node_l == NULL) {
```
I missed this catch, done now

nniranjhana commented:
@ramakrishnapallala thanks for initial review, fixed the changes requested in your comments, request to review this version

@lpereira request to review again

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -16,6 +16,10 @@
 
 #define LOWEST_POSIX_THREAD_PRIORITY 1
 
+PTHREAD_MUTEX_DEFINE(lock);
```
pthread_key_lock

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -16,6 +16,10 @@
 
+PTHREAD_MUTEX_DEFINE(lock);
+
+PTHREAD_COND_DEFINE(cvar0);
```
same here

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -302,6 +306,34 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+	}
+
+	if (*once == PTHREAD_ONCE_NOT_INIT) {
+		pthread_mutex_lock(&lock);
```
move lock out of this block and hold while reading as well.

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -302,6 +306,34 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+	}
+
+	while (*once != PTHREAD_ONCE_INIT) {
+		pthread_mutex_lock(&lock);
```
move before while and add unlock

ramakrishnapallala commented on lib/posix/pthread.c:
```diff
@@ -302,6 +306,38 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+		pthread_mutex_unlock(&pthread_key_lock);
+		return 0;
+	} else {
+		pthread_mutex_trylock(&pthread_key_lock);
```
remove this

ramakrishnapallala commented on tests/posix/pthread_key/src/pthread_key.c:
```diff
@@ -0,0 +1,95 @@
+K_THREAD_STACK_ARRAY_DEFINE(stacks, 1, STACKSZ);
+
+pthread_key_t key;
+static pthread_once_t key_once = PTHREAD_ONCE_NOT_INIT;
```
no need of explicit assignment here

nniranjhana replied:
removed it

nniranjhana commented:
@lpereira @nashif request to review

lpereira commented on lib/posix/pthread.c:
```diff
@@ -302,6 +306,37 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+		pthread_mutex_unlock(&pthread_key_lock);
+		return 0;
+	} else {
+		(*initFunc)();
```
Use the shorthand notation: `init_func()` instead of `(*init_func)()`.

lpereira commented on lib/posix/pthread.c:
```diff
@@ -302,6 +306,37 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+		return 0;
+	}
+
+	pthread_mutex_lock(&pthread_key_lock);
```
All this code from line 328 to 337 is unreachable due to the `else` condition in the previous `if` being a catchall.

nniranjhana replied:
Oh yeah! Realized I can remove the else block now, will fix this

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,179 @@
+
+	*key = NULL;
+
+	k_sem_init(&pthread_key_sem, 1, 1);
```
Why is pthread_key_sem() initialized every time pthread_key_create() is called?

nniranjhana replied:
Okay, so my understanding is that, for every new key we create, we will need to initialize a semaphore, which will be taken and given when we modify this key's (the respective key created with `pthread_key_create`) thread list using the pthread_get_specific, etc functions.

And this is where `pthread_once()` is useful - we will pass `pthread_key_create()` to `pthread_once()` so that the function is called only once for a particular key creation. So for the same key, there won't be multiple initialization of the semaphore.

lpereira replied:
There's just one semaphore for all keys, and it's being initialized every time. It's not going to work.

nniranjhana replied:
Okay, I got your point, I've statically initialized now

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,179 @@
+				 * associate thread specific data
+				 */
+				thread_spec_data->spec_data = (void *)value;
+				break;
```
Prefer an early return idiom: it reduces the mental overhead while reasoning about a piece of code.  Instead of a `break` here, use `goto out`.

nniranjhana replied:
Okay will add that

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,179 @@
+					(sys_dnode_t *)key_data);
+		}
+	}
+
```
Add `out:` here.

nniranjhana replied:
Done

lpereira commented on tests/posix/pthread_key/src/pthread_key.c:
```diff
@@ -0,0 +1,95 @@
+
+static void make_key(void)
+{
+	(void) pthread_key_create(&key, NULL);
```
Please also test the creation of multiple keys, and make sure that pthread_getspecific() and pthread_setspecific() will set/get the values for multiple keys.  Also perform negative testing (e.g. pthread_getspecific(key1) should not return any values set with pthread_setspecific(key2, ...)).

lpereira replied:
Also, tests should never cast the result of something that can fail to `(void)`.  Ensure that you can make pthread_key_create() fail and test for that.

nniranjhana replied:
Yes, I was planning to add two more functions to this test case once this patch gets merged - for multiple threads using same key, and same thread using multiple keys complete with negative tests. I will add this test later this week in M3, I will remove the cast here and check for failure for now. Hope this is okay.

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,179 @@
+	 */
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	sys_dlist_init(&thread->key_list);
```
Why is this initialized here? Calling pthread_setspecific() multiple times will reset this list.

nniranjhana replied:
That's an important catch, thanks. I have removed it here and added the thread's key list initialization during `pthread_create`.

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,179 @@
+			key_data->thread_data.spec_data = (void *)value;
+
+			/* Append new thread key data to thread's key list */
+			sys_dlist_append((sys_dlist_t *)(&thread->key_list),
```
Why is the cast to `sys_dlist_t*` needed here? `&thread->key_list` is already the right type.  

Also, there are two lists that are being maintained, and only one is taken care of when disposing of keys.  Is this correct?

nniranjhana replied:
Okay, I will remove the cast.

Yes there are two lists, and the one I am not taking care of is part of the posix thread struct, which will be disposed of along with the other thread data during deallocation.

lpereira reviewed (changes requested):
In addition to inline comments, please also fix the indentation; at least from the view in GitHub, it seems that this code is indented using 4 spaces instead of tabs as required by the coding style.

nniranjhana commented:
@lpereira thanks a lot for reviewing in detail again, I have pushed with required changes, request to review them now..

Also, I checked, I am using tabs (using vim, it automatically uses tabs when I hit enter), could you point to me where I am using spaces?

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,178 @@
+		key_data = k_malloc(sizeof(pthread_key_data));
+
+		if (key_data == NULL) {
+			return ENOMEM;
```
pthread_key_sem is locked here.  Create a `retval` variable, set it to `ENOMEM`, and `goto out`.

nniranjhana replied:
oh yeah, fixed this now

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,178 @@
+
+	/* Traverse the list of keys set by the thread, looking for key */
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(thread->key_list), node_l, next_node_l) {
```
You don't need to sue the SAFE variant of this macro if you're only traversing the list.  The SAFE variant is meant to be used when you're removing nodes as you traverse the list (as it saves the next pointer).

lpereira commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,178 @@
+	 */
+	k_sem_take(&pthread_key_sem, K_FOREVER);
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(thread->key_list), node_l, next_node_l) {
```
Same comment about using the SAFE variant.

lpereira reviewed (changes requested):
The indentation seems fine, it's just a GitHub issue here (I'm using a browser extension to fix some issues I have with it and it changes how code is presented).

Things are looking better, though, but there are a few things that need to be fixed before this can get merged

nniranjhana commented:
@lpereira Okay thanks again, and I have fixed the issues, do review once more

nniranjhana commented:
@lpereira I have added the destructor call according to the specification at the end of `pthread_exit()`, now since this makes use of `pthread_thread_data` and `pthread_key_obj`, I moved it back to the types.h file

Request to review now

nniranjhana commented:
@lpereira I have addressed all your review comments, request to review once more if there is anything further to be modified before 1.12 merge window closure

nniranjhana commented:
@andyross could you review this to see if there is anything to be modified before 1.12 merge window closure?

Commit:
tests: posix: add test for pthread_key

Added a simple test case for pthread_key, which
tests all the APIs - the thread creates a key,
sets a value to the key, retrieves and asserts
they are equal, and then deletes the key.

Signed-off-by: Niranjhana N <niranjhana.n@intel.com>

nniranjhana commented:
@nashif @andyross I have updated my test after removing the new line characters in zassert and also adding the documentation for test. This test currently tests the 5 APIs I have added. Once this patch is merged, I will push two more test cases to the test such as multiple threads using single key and negative tests, taking this as enhancement.

Request to review and merge implementation

nniranjhana commented:
Hi @lpereira and @andyross I have rebased my PR on top of latest commits, request to provide review now

nashif commented:
@nniranjhana looks like a few comments have not been addressed.

nniranjhana commented:
Hi @nashif If it is the destructor call, I have added it in `pthread_exit` function after @lpereira 's review. As his review is on `pthread_key_delete` function, it appears to be unaddressed.

nniranjhana commented on include/posix/sys/types.h:
```diff
@@ -76,6 +76,21 @@ typedef struct pthread_rwlock_obj {
+	sys_dlist_t key_data_l;
+
+	/* Optional destructor that is passed to pthread_key_create() */
+	void (*destructor)(void *);
```
I have added the destructor call in `pthread_exit()`

nniranjhana commented on include/posix/sys/types.h:
```diff
@@ -76,6 +76,21 @@ typedef struct pthread_rwlock_obj {
+/* pthread_key */
+typedef void *pthread_key_t;
+
+typedef struct pthread_key_obj {
```
In types.h file, we do have similar `pthread_rwlock_obj` implementation detail, and I am not sure where else I would be adding this, since at least two other functions (pthread_key APIs and pthread_exit) are making use of `pthread_thread_data` and `pthread_key_obj`; and so I cannot add it to the pthread_key.c file. So could you help me here? @lpereira @nashif

lpereira replied:
You can create header files that are included only by the implementation files.  These don't need to be in user-exposed headers.

nniranjhana replied:
Okay sure @lpereira I will do that, thanks

andyross commented on lib/posix/pthread.c:
```diff
@@ -299,6 +303,28 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
+ */
+int pthread_once(pthread_once_t *once, void (*init_func)(void))
+{
+	pthread_mutex_lock(&pthread_key_lock);
```
Why use a pthread lock here instead of a semaphore?  You're using a k_sem below for a lock...

nniranjhana replied:
No particular reason other than that most `pthread` functions in the pthread.c file used `pthread_mutex_lock` functions (like pthread_cancel, pthread_create, etc.). So in keeping with that, I used this for pthread_once and the addition in pthread_exit

andyross commented on lib/posix/pthread_key.c:
```diff
@@ -0,0 +1,165 @@
+	pthread_key_data *key_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	k_sem_take(&pthread_key_sem, K_FOREVER);
```
Conversely, why use a k_sem here instead of a pthread mutex, which you're using above for the same kind of lock?

nniranjhana replied:
When I took a look at the specific implementation of POSIX functions like message queues, pthread_rwlock, etc, they directly made use of k_sem objects, and so I followed the same for pthread_key.c

andyross commented on include/posix/pthread.h:
```diff
@@ -30,6 +31,9 @@ enum pthread_state {
 	struct k_thread thread;
 
+	/* List of keys that thread has called pthread_setspecific() on */
+	sys_dlist_t key_list;
```
I don't see why this has to be a dlist.  You don't ever do anything but search the list linearly from front to back, and only remove elements having found them via linear search.  This should be doable with an slist, no?

nniranjhana replied:
Ah, good catch. I'm not sure why I used dlist in the beginning. I see the singly linked list implementation is circular too. You are right, I will make use of `slist` only to save space.

andyross commented on include/posix/pthread_key.h:
```diff
@@ -0,0 +1,52 @@
+	/* List of pthread_key_data objects that contain thread
+	 * specific data for the key
+	 */
+	sys_dlist_t key_data_l;
```
Same comment: this could be an slist and save a word of storage in every struct.

andyross reviewed (approved):
Mostly nitpicking.  Can't see anything that should block merge.

nniranjhana commented:
@andyross Thanks a lot for reviewing this, I have addressed your review comments, and replaced dlist with slist. Is there anything else?
@nashif Request to take a look for merging this, thanks

nashif commented on tests/posix/pthread_key/testcase.yaml:
```diff
@@ -0,0 +1,3 @@
+tests:
+  posix.pthread_key:
```
portability.posix.pthread_key

nashif commented on tests/posix/pthread_key/testcase.yaml:
```diff
@@ -0,0 +1,3 @@
+tests:
+  posix.pthread_key:
+    tags: core
```
tags: posix
